### PR TITLE
Windowed time decay functions

### PR DESCRIPTION
**Pull request by excitoon**
Changelog category (leave one):
- New Feature

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Introduced window functions:
- `exponentialTimeDecayedSum`
- `exponentialTimeDecayedMax`
- `exponentialTimeDecayedCount`
- `exponentialTimeDecayedAvg`
which are more effective than `exponentialMovingAverage` for bigger windows. Also more use-cases were covered.

### CODE REVIEW

**Comment by excitoon** [comment]
Related #28914

**Comment by excitoon** [comment]
```
SELECT number%100 = 0 AS value, number AS time, exponentialMovingAverage(1)(value, time) OVER (ROWS BETWEEN 100 PRECEDING AND CURRENT ROW) AS exp_smooth FROM numbers(1000000);
...
1000000 rows in set. Elapsed: 1.986 sec. Processed 1.05 million rows, 8.38 MB (527.84 thousand rows/s., 4.22 MB/s.)
```

```
SELECT number%100 = 0 AS value, number AS time, exponentialTimeDecayedSum(1/ln(2))(value, time) OVER (ROWS BETWEEN 100 PRECEDING AND CURRENT ROW) AS exp_smooth FROM numbers(1000000);
...
1000000 rows in set. Elapsed: 0.096 sec. Processed 1.05 million rows, 8.38 MB (10.97 million rows/s., 87.77 MB/s.)
```

**Comment by alexey-milovidov** [comment]
`SELECT NULL, exponentialTimeDecayedSum(100000002004087730000.)(value, time) OVER (Rows BETWEEN 255 FOLLOWING AND UNBOUNDED FOLLOWING), number = -2147483649, exponentialTimeDecayedSum(-1)(value, time) OVER (Rows BETWEEN UNBOUNDED PRECEDING AND 65537 PRECEDING), number AS value, number = NULL, exponentialTimeDecayedSum(65537)(value, time) OVER (Rows BETWEEN UNBOUNDED PRECEDING AND 1048577 FOLLOWING), number AS time FROM numbers(65535)`

Segfault.

**Comment by excitoon** [comment]
Okay, there is something about window functions I do not know about so far.

**Review by mathalex — COMMENTED** [review]
The code has a lot of copy-paste. Especially `Sum`, `Max`, `Avg` functions. So I recommend making constructors and  `windowInsertResultInto` function more abstract.

File: src/Processors/Transforms/WindowTransform.cpp [hunk 1]
```diff
@@ -1513,6 +1514,340 @@ struct WindowFunctionDenseRank final : public WindowFunction
...
+
+        Float64 last_val = getLastValueFromOutputColumn<Float64>(transform, function_index);
+        Float64 last_t = getLastValueFromInputColumn<Float64>(transform, function_index, ARGUMENT_TIME);
+
+        IColumn & to = *current_block.output_columns[function_index];
+
+        Float64 x = (*current_block.input_columns[workspace.argument_column_indices[ARGUMENT_VALUE]]).getFloat64(transform->current_row.row);
+        Float64 t = (*current_block.input_columns[workspace.argument_column_indices[ARGUMENT_TIME]]).getFloat64(transform->current_row.row);
+
+        Float64 c = exp((last_t-t)/decay_length);
```
[the remark this answers is not in the document]
Comment (excitoon) [reply src/Processors/Transforms/WindowTransform.cpp]: +

**Comment by alexey-milovidov** [comment]
Let's fix the fast test and let @mathalex review it.

**Comment by alexey-milovidov** [comment]
@mergifyio update

**Review by mathalex — COMMENTED** [review]
Overall, the code is OK, with clear logic. Maybe some performance tests can be added to prove `Introduced window functions ... which are more effective than exponentialMovingAverage for bigger windows` statement. Especially I liked the implementation of `WindowFunctionExponentialTimeDecayedAvg`. Also there is a way to use not sum and count, but avg and count (so one of state fields is a result), but recalculations are a bit harder.

File: src/Processors/Transforms/WindowTransform.cpp [hunk 2]
```diff
@@ -1513,6 +1507,410 @@ struct WindowFunctionDenseRank final : public WindowFunction
     }
 };
 
+namespace recurrent_detail
+{
+    template<typename T> T getLastValueFromInputColumn(const WindowTransform * /*transform*/, size_t /*function_index*/, size_t /*column_index*/)
+    {
+        throw Exception(ErrorCodes::NOT_IMPLEMENTED, "getLastValueFromInputColumn() is not implemented");
```
[the remark this answers is not in the document]
Comment (excitoon) [reply src/Processors/Transforms/WindowTransform.cpp]: +

**Review by nikitamikhaylov — APPROVED** [review]

**Review by nikitamikhaylov — APPROVED** [review]
Ok, but add comments about all the functions. And why do we need template helpers such as getLastValueFromInputColumn, but with only one specialization for Float64?

**Comment by nikitamikhaylov** [comment]
> Tests are OK. Merge? @nikitamikhaylov

Exactly, but after adding comments. A lack of comments contradicts with the rest code of window function.

**Review by mathalex — APPROVED** [review]
Function names are self-explaining, so you can add comments just for the sake of code consistency.

**Comment by mathalex** [comment]
> And why do we need template helpers such as getLastValueFromInputColumn, but with only one specialization for Float64?

I think these helpers are quite useful, therefore template implementation is used so that you can easily rewrite these functions for other types in the future.

**Comment by sevirov** [comment]
Internal documentation ticket: DOCSUP-19039

**Comment by rfraposa** [comment]
@nikitamikhaylov - It's over 2 years ago and these new functions are not in the docs. Do they work? Should we add them to the docs and mention that they might be more effective over larger windows?

**Comment by nikitamikhaylov** [comment]
@rfraposa At least they should work, I didn't see corresponding tests being failed. I'm not sure they actually more effective, I didn't read the code but maybe (better not to mention this). Of course we need to add them to docs. Thanks in advance if you can do this 🙏

**Pull request merged**